File: test-suite/handwritten-src/java/com/dropbox/djinni/test/DataTest.java
package com.dropbox.djinni.test;
import junit.framework.TestCase;
import static org.junit.Assert.*;
import java.nio.ByteBuffer;

public class DataTest extends TestCase {
    DataRefTest test;

    @Override
    protected void setUp() {
        test = DataRefTest.create();
    }

    public void testSend() {
        byte[] input = new byte[]{0, 1, 2, 3};
        ByteBuffer buf = ByteBuffer.allocateDirect(4);
        buf.put(input);
        test.sendData(buf);
        byte[] output = test.retriveAsBin();
        assertArrayEquals(input, output);
    }

    public void testSendMutable() {
        byte[] input = new byte[]{0, 1, 2, 3};
        ByteBuffer buf = ByteBuffer.allocateDirect(4);
        buf.put(input);
        test.sendMutableData(buf);
        buf.rewind();
        byte[] output = new byte[4];
        buf.get(output);
        assertArrayEquals(new byte[]{3, 2, 1, 0}, output);
    }

    public void testGenerateData() {
        ByteBuffer buf = test.generateData();
        byte[] output = new byte[4];
        buf.get(output);
        assertArrayEquals(new byte[]{0, 1, 2, 3}, output);
    };

    public void testDataFromVec() {
        ByteBuffer buf = test.dataFromVec();
        byte[] output = new byte[4];
        buf.get(output);
        assertArrayEquals(new byte[]{0, 1, 2, 3}, output);
    }

    public void testDataFromStr() {
        ByteBuffer buf = test.dataFromStr();
        byte[] output = new byte[4];
        buf.get(output);
        assertArrayEquals(new byte[]{'a', 'b', 'c', 'd'}, output);
    }

    public void testSendDataView() {
        byte[] input = new byte[]{0, 1, 2, 3};
        ByteBuffer buf = ByteBuffer.allocateDirect(4);
        buf.put(input);
        byte[] output = test.sendDataView(buf);
        assertArrayEquals(input, output);
    }

    public void testRecvDataView() {
        ByteBuffer buf = test.recvDataView();
        byte[] output = new byte[4];
        buf.get(output);
        assertArrayEquals(new byte[]{0, 1, 2, 3}, output);
    }
}

File: test-suite/handwritten-src/wasm/DataTest.js
class DataTest {
    constructor(module) {
        this.m = module;
    }

    setUp() {
        this.test = this.m.DataRefTest.create();
    }

    testSend() {
        var input = [0, 1, 2, 3];
        var buf = this.m.allocateDirectBuffer(4);
        buf.set(input);
        this.test.sendData(buf);
        var output = this.test.retriveAsBin();
        assertArrayEq(input, output);
    }

    testSendMutable() {
        var input = [0, 1, 2, 3];
        var buf = this.m.allocateDirectBuffer(4);
        buf.set(input);
        this.test.sendMutableData(buf);
        assertArrayEq([3, 2, 1, 0], buf);
    }

    testGenerateData() {
        var buf = this.test.generateData();
        assertArrayEq([0, 1, 2, 3], buf);
    };

    testDataFromVec() {
        var buf = this.test.dataFromVec();
        assertArrayEq([0, 1, 2, 3], buf);
    }

    testDataFromStr() {
        var buf = this.test.dataFromStr();
        // codepoints of a, b, c, d
        assertArrayEq([97, 98, 99, 100], buf);
    }

    testSendDataView() {
        var input = [0, 1, 2, 3];
        var buf = this.m.allocateDirectBuffer(4);
        buf.set(input);
        var output = this.test.sendDataView(buf);
        assertArrayEq(input, output);
    }
    
    testRecvDataView() {
        var output = this.test.recvDataView();
        assertArrayEq([0, 1, 2, 3], output);
    }
}

allTests.push(DataTest);
